### PR TITLE
ddl: support column type change from string to other types

### PR DESCRIPTION
**Pull request by LENSHOOD**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/19971

Problem Summary:
DDL modify column, change type from different types, see more detail at: https://github.com/pingcap/tidb/issues/19939

### What is changed and how it works?
What's Changed:
Support ddl column type change from string to other types.

According [MySQL data types](https://dev.mysql.com/doc/refman/8.0/en/string-types.html), string types include char/varchar/binary/varbinary/blob/text/enum/set, most of them can directly cast to other types(see correlated integration test at column_type_change_test.go).

However, there are two type cannot cast yet:
- enum/set to date time type: still no cast logic at current revision, enum/set to date time not supported
- string to bit: the behavior not compatible with MySQL

Above two cases should fix at individual PRs (will create issues later)

### Check List 

Tests 

- Integration test

### Release note 

- ddl: support column type change from string to other types

### CODE REVIEW

**Comment by LENSHOOD** [comment]
@ti-srebot /run-all-tests

**Comment by LENSHOOD** [comment]
@AilinKid PTAL

**Comment by AilinKid** [comment]
> @AilinKid PTAL

Good for it!

File: ddl/column.go [hunk 1]
```diff
@@ -647,36 +648,39 @@ func onSetDefaultValue(t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
+	needTruncationOrToggleSign := func() bool {
+		return newCol.Flen > 0 && newCol.Flen < oldCol.Flen || toUnsigned != originUnsigned
 	}
 
-	switch oldCol.Tp {
-	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
-		switch newCol.Tp {
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
-			return oldCol.Tp != newCol.Tp
+	// deal with same type
```
Comment (AilinKid) [inline ddl/column.go:655]:
```suggestion
	// Deal with the same type.
```
the below comment should also be paid with attention.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3361,10 +3361,11 @@ func CheckModifyTypeCompatible(origin *types.FieldType, to *types.FieldType) (al
 			mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
 			skipSignCheck = true
 			skipLenCheck = true
-		case mysql.TypeEnum, mysql.TypeSet:
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		default:
+		case mysql.TypeBit:
+			// todo: currently string data type cast to bit are not compatible with mysql, should fix here after compatible
```
Comment (AilinKid) [inline ddl/ddl_api.go:3365]: Do you mean the bit length is required more to change in TiDB?

File: ddl/column_type_change_test.go [hunk 3]
```diff
@@ -444,3 +444,225 @@ func (s *testColumnTypeChangeSuite) TestColumnTypeChangeFromIntegerToOthers(c *C
...
+	}
+
+	// wrong json data
+	reset(tk)
+	tk.MustExec("insert into t values ('{\"k1\": \"value\"')")
+	tk.MustGetErrCode("alter table t modify vc json", mysql.ErrInvalidJSONText)
+	// decimal bad number
+	reset(tk)
+	tk.MustExec("insert into t values ('abc')")
+	tk.MustGetErrCode("alter table t modify vc decimal(5,3)", mysql.ErrBadNumber)
```
Comment (AilinKid) [inline ddl/column_type_change_test.go:667]: there are some cases be treated as an error while as a warning in TiDB.
for example:
"1a" will be truncated as 1 with waring in TiDB, but it errors in MySQL, this kind of incompatibility is acceptable, and should list in the test here.

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: ddl/column_type_change_test.go (hunk 3 above)
Comment (LENSHOOD) [reply ddl/column_type_change_test.go:667]: sure, I'll add tests for them.

File: ddl/ddl_api.go (hunk 2 above)
Comment (LENSHOOD) [reply ddl/ddl_api.go:3365]: Yes, the Datum.convertToMysqlBit() will treat "1" as ascii 0x31, then lead to data truncate error. But MySQL can cast "1" to 0x01. 

Maybe it's better to create an individual issue to solve it.

Comment (AilinKid) [reply ddl/ddl_api.go:3365]: Great

Comment (LENSHOOD) [reply ddl/ddl_api.go:3365]: Got it.

**Review by djshow832 — COMMENTED** [review]
LGTM

**Review by AilinKid — COMMENTED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

**Comment by LENSHOOD** [comment]
@ti-srebot /run-all-tests

**Pull request merged**